src/Swaps/Swapper.sol

```diff
@@ -0,0 +1,134 @@
+
+import {ISwapper} from "./ISwapper.sol";
+
+import {OneInchHandler} from "./handlers/OneInchHandler.sol";
+import {UniswapV2Handler} from "./handlers/UniswapV2Handler.sol";
+import {UniswapV3Handler} from "./handlers/UniswapV3Handler.sol";
+import {UniswapAutoRouterHandler} from "./handlers/UniswapAutoRouterHandler.sol";
+
+/// @title Swapper
+/// @custom:security-contact security@euler.xyz
+/// @author Euler Labs (https://www.eulerlabs.com/)
+/// @notice Untrusted helper contract for EVK for performing swaps and swaps to repay
+contract Swapper is OneInchHandler, UniswapV2Handler, UniswapV3Handler, UniswapAutoRouterHandler {
+    bytes32 public constant HANDLER_ONE_INCH = bytes32("1Inch");
+    bytes32 public constant HANDLER_UNISWAP_V2 = bytes32("UniswapV2");
+    bytes32 public constant HANDLER_UNISWAP_V3 = bytes32("UniswapV3");
+    bytes32 public constant HANDLER_UNISWAP_AUTOROUTER = bytes32("UniswapAutoRouter");
+
+    uint256 internal constant REENTRANCYLOCK_UNLOCKED = 1;
+    uint256 internal constant REENTRANCYLOCK_LOCKED = 2;
+
+    uint256 private reentrancyLock;
+
+    error Swapper_UnknownMode();
+    error Swapper_UnknownHandler();
+    error Swapper_Reentrancy();
+    error Swapper_InsufficientBalance();
+    error Swapper_PastDeadline();
+
+    // In the locked state, allow contract to call itself, but block all external calls
+    modifier externalLock() {
+        bool isExternal = msg.sender != address(this);
+
+        if (isExternal) {
+            if (reentrancyLock == REENTRANCYLOCK_LOCKED) revert Swapper_Reentrancy();
+            reentrancyLock = REENTRANCYLOCK_LOCKED;
+        }
+
+        _;
+
+        if (isExternal) reentrancyLock = REENTRANCYLOCK_UNLOCKED;
+    }
+
+    constructor(address oneInchAggregator, address uniswapRouterV2, address uniswapRouterV3, address uniswapRouter02)
+        OneInchHandler(oneInchAggregator)
+        UniswapV2Handler(uniswapRouterV2)
+        UniswapV3Handler(uniswapRouterV3)
+        UniswapAutoRouterHandler(uniswapRouter02)
+    {}
+
+    /// @inheritdoc ISwapper
+    function swap(SwapParams memory params)
+        public
+        externalLock
+    {
+        if (params.mode >= MODE_MAX_VALUE) revert Swapper_UnknownMode();
+        if (params.deadline < block.timestamp) revert Swapper_PastDeadline();
+
+        if (params.handler == HANDLER_ONE_INCH) {
+            swapOneInch(params);
+        } else if (params.handler == HANDLER_UNISWAP_V2) {
+            swapUniswapV2(params);
+        } else if (params.handler == HANDLER_UNISWAP_V3) {
+            swapUniswapV3(params);
+        } else if (params.handler == HANDLER_UNISWAP_AUTOROUTER) {
+            swapAutoRouter(params);
+        } else {
+            revert Swapper_UnknownHandler();
+        }
+
+        if (params.mode == MODE_EXACT_IN) return;
+
+        // swapping to target debt is only useful for repaying
+        if (params.mode == MODE_TARGET_DEBT) {
```

Comment:
I know it would be slightly more expensive, but wouldn't it make more sense to have the following modes?
- exact in
- exact out
- exact out target debt

It would mean that the `exact out target debt` doesn't need to repay anything, it will only swap into the swapper and then one would need to call `repayAndDeposit` themselves as an additional multicall action. I think it would simplify the reasoning about this contract and possibily simplify the `SwapParams` that now is full of exceptions depending on the mode.

Reply:
if we split it up, then we wouldn't know how much to repay, unless we repeat the target calculation at a cost. We can't just assume we have exact amount for repay after the swap, because we could have overswapped or we were griefed bo someone sending output tokens to the swapper